3 course/OperatingSystem/mythread.c

```diff
@@ -0,0 +1,173 @@
+#define _GNU_SOURCE
+#include <stdlib.h>
+#include <stdio.h>
+#include <pthread.h>
+#include <unistd.h>
+#include <sys/syscall.h>
+#include <linux/futex.h>
+#include <stdatomic.h>
+#include <sched.h>
+
+#include <sys/mman.h>
+#include <string.h>
+#include <errno.h>
+#include <sys/time.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <stdint.h>
+
+#include <signal.h>
+#include <pthread.h>
+
+#define PAGE 4096
+#define STACK_SIZE PAGE*8
+
+typedef void *(*start_routine_t)(void*);
+
+typedef struct _mythread {
+    int mythread_id;
+    start_routine_t start_routine;
+    void* arg;
+    void* retval;
+    volatile int joined;
+    volatile int exited;
+
+} mythread_struct_t;
+
+void* create_stack(off_t size, int thread_id) {
+  void* stack;
+
+  stack = mmap(NULL, size, PROT_READ | PROT_WRITE, MAP_STACK | MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+  if (!stack) {
+    printf("Error: stack = NULL\n");
+    return NULL;
+  }
+
+  printf("Stack ptr#: %p\n", stack);
+
+  return stack;
+}
+
+void wait_mythread(volatile int* lock) {
+    const int one = 1;
+    int err;
+
+    while (1) {
+        if (atomic_compare_exchange_strong(lock, &one, 1)) {
+            return;
+        }
+
+        err = syscall(SYS_futex, &lock, FUTEX_WAIT, 0, NULL, NULL, 0);
+        if (err == -1 && errno != EAGAIN) {
+            printf("Futex FUTEX_WAIT failed: %s\n", strerror(errno));
+            abort();
+        }
+    }
+}
+
+void wake_mythread(volatile int* lock) {
+    const int zero = 0;
+    int err; 
+   
+    if (atomic_compare_exchange_strong(lock, &zero, 1)) {
+        err = syscall(SYS_futex, &lock, FUTEX_WAKE, 1, NULL, NULL, 0);
+        if (err == -1) {
+            printf("Futex FUTEX_WAKE failed: %s\n", strerror(errno));
+            abort();
+        }
+    }
+}
+
+int mythread_startup(void* arg) {
+    mythread_struct_t* mythread = (mythread_struct_t*)arg;
+
+    mythread->retval = mythread ->start_routine(mythread->arg);
+    //mythread->exited = 1;
+    //меняем значение в функции ниже
+    wake_mythread(&mythread->exited);
+
+    //wait until join (тот самый futex!)
+    wait_mythread(&mythread->joined);
+
+    return 0;
+}
+
+void destroy(mythread_struct_t* mythread) {
+    munmap((void*)mythread, STACK_SIZE);
+}
+
+int mythread_join(uintptr_t mytid, void** retval) {
```

Comment:
Please explain what happen if several threads will wait for the same thread .

Reply:
Memory is cleared twice for this thread

Reply:
why it is not fixed ?